java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Duplicate field added thrown during JUNIT test

### Describe the bug

I am upgrading quarkus from 3.29.4 to 3.30.1 and i encounter the following error during the junit test:


java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.reflect.UndeclaredThrowableException
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:878)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Duplicate field added: com.nn.aav.insurance.screening.business.inquiries.boundary.InquiriesAPILogger_pUyQ2LSzOfLd-I4OpCNbAtvXr44_Delegate_Subclass#sLGN0e0eUKbgl46r1-BV6f4mZ88:java.lang.Object
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:443)
	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:538)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:872)
	... 8 more
Caused by: java.lang.IllegalArgumentException: Duplicate field added: com.nn.aav.insurance.screening.business.inquiries.boundary.InquiriesAPILogger_pUyQ2LSzOfLd-I4OpCNbAtvXr44_Delegate_Subclass#sLGN0e0eUKbgl46r1-BV6f4mZ88:java.lang.Object
	at io.quarkus.gizmo2.impl.ClassCreatorImpl.field(ClassCreatorImpl.java:67)
	at io.quarkus.arc.processor.SubclassGenerator.lambda$processDecorator$38(SubclassGenerator.java:641)
	at io.quarkus.gizmo2.impl.GizmoImpl.lambda$class_$1(GizmoImpl.java:154)
	at io.github.dmlloyd.classfile.impl.ClassFileImpl.build(ClassFileImpl.java:144)
	at io.github.dmlloyd.classfile.ClassFile.build(ClassFile.java:558)
	at io.quarkus.gizmo2.impl.GizmoImpl.class_(GizmoImpl.java:151)
	at io.quarkus.gizmo2.Gizmo.class_(Gizmo.java:89)
	at io.quarkus.arc.processor.SubclassGenerator.processDecorator(SubclassGenerator.java:621)
	at io.quarkus.arc.processor.SubclassGenerator.lambda$createSubclass$7(SubclassGenerator.java:236)
	at io.quarkus.gizmo2.impl.ConstructorCreatorImpl.lambda$body$0(ConstructorCreatorImpl.java:58)
	at io.quarkus.gizmo2.impl.BlockCreatorImpl.accept(BlockCreatorImpl.java:992)
	at io.quarkus.gizmo2.impl.ExecutableCreatorImpl.doCode(ExecutableCreatorImpl.java:405)
	at io.quarkus.gizmo2.impl.ExecutableCreatorImpl.lambda$doBody$1(ExecutableCreatorImpl.java:317)
	at io.github.dmlloyd.classfile.impl.DirectCodeBuilder.build(DirectCodeBuilder.java:93)
	at io.github.dmlloyd.classfile.impl.DirectMethodBuilder.withCode(DirectMethodBuilder.java:118)
	at io.github.dmlloyd.classfile.impl.DirectMethodBuilder.withCode(DirectMethodBuilder.java:125)
	at io.quarkus.gizmo2.impl.ExecutableCreatorImpl.doBody(ExecutableCreatorImpl.java:305)
	at io.quarkus.gizmo2.impl.ExecutableCreatorImpl.lambda$body$3(ExecutableCreatorImpl.java:436)
	at io.github.dmlloyd.classfile.impl.DirectMethodBuilder.run(DirectMethodBuilder.java:139)
	at io.github.dmlloyd.classfile.impl.DirectClassBuilder.withMethod(DirectClassBuilder.java:117)
	at io.github.dmlloyd.classfile.ClassBuilder.withMethod(ClassBuilder.java:313)
	at io.quarkus.gizmo2.impl.ExecutableCreatorImpl.body(ExecutableCreatorImpl.java:435)
	at io.quarkus.gizmo2.impl.ConstructorCreatorImpl.body(ConstructorCreatorImpl.java:53)
	at io.quarkus.arc.processor.SubclassGenerator.lambda$createSubclass$8(SubclassGenerator.java:208)
	at io.quarkus.gizmo2.impl.ConstructorCreatorImpl.accept(ConstructorCreatorImpl.java:63)
	at io.quarkus.gizmo2.impl.ClassCreatorImpl.constructor(ClassCreatorImpl.java:118)
	at io.quarkus.arc.processor.SubclassGenerator.lambda$createSubclass$17(SubclassGenerator.java:185)
	at io.quarkus.gizmo2.impl.GizmoImpl.lambda$class_$1(GizmoImpl.java:154)
	at io.github.dmlloyd.classfile.impl.ClassFileImpl.build(ClassFileImpl.java:144)
	at io.github.dmlloyd.classfile.ClassFile.build(ClassFile.java:558)
	at io.quarkus.gizmo2.impl.GizmoImpl.class_(GizmoImpl.java:151)
	at io.quarkus.gizmo2.Gizmo.class_(Gizmo.java:89)
	at io.quarkus.arc.processor.SubclassGenerator.createSubclass(SubclassGenerator.java:130)
	at io.quarkus.arc.processor.SubclassGenerator.generate(SubclassGenerator.java:120)
	at io.quarkus.arc.processor.BeanProcessor$4$2.call(BeanProcessor.java:358)
	at io.quarkus.arc.processor.BeanProcessor$4$2.call(BeanProcessor.java:355)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at --- Async.Stack.Trace --- (captured by IntelliJ IDEA debugger)
	at java.base/java.util.concurrent.FutureTask.<init>(FutureTask.java:132)
	at java.base/java.util.concurrent.AbstractExecutorService.newTaskFor(AbstractExecutorService.java:113)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:144)
	at io.quarkus.arc.processor.BeanProcessor$4.call(BeanProcessor.java:355)
	at io.quarkus.arc.processor.BeanProcessor$4.call(BeanProcessor.java:323)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at --- Async.Stack.Trace --- (captured by IntelliJ IDEA debugger)
	at java.base/java.util.concurrent.FutureTask.<init>(FutureTask.java:132)
	at java.base/java.util.concurrent.AbstractExecutorService.newTaskFor(AbstractExecutorService.java:113)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:144)
	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:323)
	... 11 more

	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:388)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:305)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:61)
	at io.quarkus.test.junit.AppMakerHelper.getStartupAction(AppMakerHelper.java:253)
	at io.quarkus.test.junit.classloading.FacadeClassLoader.getOrCreateRuntimeClassLoader(FacadeClassLoader.java:587)
	at io.quarkus.test.junit.classloading.FacadeClassLoader.getQuarkusClassLoader(FacadeClassLoader.java:484)
	at io.quarkus.test.junit.classloading.FacadeClassLoader.loadClass(FacadeClassLoader.java:366)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:536)
	at java.base/java.lang.Class.forName(Class.java:515)
	at org.junit.platform.commons.util.ReflectionUtils.lambda$tryToLoadClass$10(ReflectionUtils.java:886)
	at org.junit.platform.commons.function.Try.lambda$call$0(Try.java:57)
	at org.junit.platform.commons.function.Try.of(Try.java:93)
	at org.junit.platform.commons.function.Try.call(Try.java:57)
	at org.junit.platform.commons.util.ReflectionUtils.tryToLoadClass(ReflectionUtils.java:873)
	at org.junit.platform.commons.util.ReflectionUtils.tryToLoadClass(ReflectionUtils.java:810)
	at org.junit.platform.commons.support.ReflectionSupport.tryToLoadClass(ReflectionSupport.java:94)
	at org.junit.platform.engine.discovery.ClassSelector.getJavaClass(ClassSelector.java:95)
	at org.junit.jupiter.engine.discovery.ClassSelectorResolver.resolve(ClassSelectorResolver.java:87)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.lambda$resolve$2(EngineDiscoveryRequestResolution.java:135)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1685)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolve(EngineDiscoveryRequestResolution.java:189)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolve(EngineDiscoveryRequestResolution.java:126)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDiscoveryRequestResolution.java:92)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.run(EngineDiscoveryRequestResolution.java:83)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolver.resolve(EngineDiscoveryRequestResolver.java:148)
	at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:63)
	at org.junit.jupiter.engine.JupiterTestEngine.discover(JupiterTestEngine.java:70)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:195)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:174)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:119)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:84)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:104)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:91)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
	at org.junit.platform.launcher.core.InterceptingLauncher.lambda$execute$1(InterceptingLauncher.java:39)
	at org.junit.platform.launcher.core.ClasspathAlignmentCheckingLauncherInterceptor.intercept(ClasspathAlignmentCheckingLauncherInterceptor.java:25)
	at org.junit.platform.launcher.core.InterceptingLauncher.execute(InterceptingLauncher.java:38)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:66)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:231)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.reflect.UndeclaredThrowableException
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:878)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Duplicate field added: com.nn.aav.insurance.screening.business.inquiries.boundary.InquiriesAPILogger_pUyQ2LSzOfLd-I4OpCNbAtvXr44_Delegate_Subclass#sLGN0e0eUKbgl46r1-BV6f4mZ88:java.lang.Object
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at io.quarkus.arc.processor.BeanProcessor.generateResources(BeanProcessor.java:443)
	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:538)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:872)
	... 8 more
Caused by: java.lang.IllegalArgumentException: Duplicate field added: com.nn.aav.insurance.screening.business.inquiries.boundary.InquiriesAPILogger_pUyQ2LSzOfLd-I4OpCNbAtvXr44_Delegate_Subclass#sLGN0e0eUKbgl46r1-BV6f4mZ88:java.lang.Object
	at io.quarkus.gizmo2.impl.ClassCreatorImpl.field(ClassCreatorImpl.java:67)
	at io.quarkus.arc.processor.SubclassGenerator.lambda$processDecorator$38(SubclassGenerator.java:641)
	at io.quarkus.gizmo2.impl.GizmoImpl.lambda$class_$1(GizmoImpl.java:154)
	at io.github.dmlloyd.classfile.impl.ClassFileImpl.build(ClassFileImpl.java:144)
	at io.github.dmlloyd.classfile.ClassFile.build(ClassFile.java:558)
	at io.quarkus.gizmo2.impl.GizmoImpl.class_(GizmoImpl.java:151)
	at io.quarkus.gizmo2.Gizmo.class_(Gizmo.java:89)
	at io.quarkus.arc.processor.SubclassGenerator.processDecorator(SubclassGenerator.java:621)

### Expected behavior

no exception!

### Actual behavior

see description

### How to Reproduce?

I am using a decorator pattern around my REST API service like this. 

```
public interface XAPI{
 
  Obvject test();

}
@Dependent
@Priority(10)
@Decorator
public class APILogger implements XAPI{

    @Inject
    @Delegate
    RealAPI resource;

}

```

I think 3.30.1 cannot handle this at this moment!
 

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.26100.7171]

### Output of `java -version`

Microsoft Windows [Version 10.0.26100.7171]

### Quarkus version or git rev

3.30.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

maven 3.9.11

### Additional information

_No response_

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
@@ -603,10 +603,11 @@ public class SubclassGenerator extends AbstractGenerator {
         }
 
         Map<MethodDesc, DecoratorMethod> nextDecorators = bean.getNextDecorators(decorator);
-        List<DecoratorInfo> decoratorParameters = new ArrayList<>();
+        Set<DecoratorInfo> decoratorParametersSet = new HashSet<>();
         for (DecoratorMethod decoratorMethod : nextDecorators.values()) {
-            decoratorParameters.add(decoratorMethod.decorator);
+            decoratorParametersSet.add(decoratorMethod.decorator);
         }
+        List<DecoratorInfo> decoratorParameters = new ArrayList<>(decoratorParametersSet);
         Collections.sort(decoratorParameters);
 
         List<ClassDesc> delegateSubclassCtorParams = new ArrayList<>();
@@ -800,19 +801,19 @@ public class SubclassGenerator extends AbstractGenerator {
         LocalVar cc = subclassCtor.localVar("cc", subclassCtor.invokeStatic(MethodDescs.CREATIONAL_CTX_CHILD, ccParam));
 
         // Create new delegate subclass instance
-        Expr[] params = new Expr[1 + decoratorParameters.size()];
-        params[0] = subclass.this_();
-        int paramIdx = 1;
+        Expr[] args = new Expr[1 + decoratorParameters.size()];
+        args[0] = subclass.this_();
+        int argIndex = 1;
         for (DecoratorInfo decoratorParameter : decoratorParameters) {
             LocalVar decoratorVar = decoratorToLocalVar.get(decoratorParameter.getIdentifier());
             if (decoratorVar == null) {
                 throw new IllegalStateException("Unknown next " + decoratorParameter + " when generating " + generatedName);
             }
-            params[paramIdx] = decoratorVar;
-            paramIdx++;
+            args[argIndex] = decoratorVar;
+            argIndex++;
         }
         Expr delegateSubclassInstance = subclassCtor.new_(ConstructorDesc.of(
-                delegateSubclass, delegateSubclassCtorParams), params);
+                delegateSubclass, delegateSubclassCtorParams), args);
 
         // Set the DecoratorDelegateProvider to satisfy the delegate IP
         LocalVar prev = subclassCtor.localVar("prev", subclassCtor.invokeStatic(

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/SameDecoratorInMultipleNextChainsTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/decorators/SameDecoratorInMultipleNextChainsTest.java
@@ -1,0 +1,90 @@
+package io.quarkus.arc.test.decorators;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import jakarta.annotation.Priority;
+import jakarta.decorator.Decorator;
+import jakarta.decorator.Delegate;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+
+/**
+ * Verifies correct behavior in case the same decorator occurs multiple times
+ * in the {@code BeanInfo.getNextDecorators()} map. The {@code SubclassGenerator}
+ * code generator didn't bother checking, which was fine in Gizmo 1 which didn't
+ * check duplicate fields. Gizmo 2 does check, so the {@code SubclassGenerator}
+ * has to guard against that.
+ */
+public class SameDecoratorInMultipleNextChainsTest {
+    @RegisterExtension
+    public ArcTestContainer container = new ArcTestContainer(MyInterface.class, MyClass.class, MyDecorator1.class,
+            MyDecorator2.class);
+
+    @Test
+    public void test() {
+        MyInterface instance = Arc.container().instance(MyInterface.class).get();
+        assertEquals("d1m1 d2m1 A", instance.method1());
+        assertEquals("d1m2 d2m2 B", instance.method2());
+    }
+
+    interface MyInterface {
+        String method1();
+
+        String method2();
+    }
+
+    @ApplicationScoped
+    static class MyClass implements MyInterface {
+        @Override
+        public String method1() {
+            return "A";
+        }
+
+        @Override
+        public String method2() {
+            return "B";
+        }
+    }
+
+    @Priority(10)
+    @Decorator
+    static class MyDecorator1 implements MyInterface {
+        @Inject
+        @Delegate
+        MyInterface resource;
+
+        @Override
+        public String method1() {
+            return "d1m1 " + resource.method1();
+        }
+
+        @Override
+        public String method2() {
+            return "d1m2 " + resource.method2();
+        }
+    }
+
+    @Priority(20)
+    @Decorator
+    static class MyDecorator2 implements MyInterface {
+        @Inject
+        @Delegate
+        MyInterface resource;
+
+        @Override
+        public String method1() {
+            return "d2m1 " + resource.method1();
+        }
+
+        @Override
+        public String method2() {
+            return "d2m2 " + resource.method2();
+        }
+    }
+}

